accord-core/src/main/java/accord/messages/ReadData.java

```diff
@@ -149,14 +149,16 @@ void read(SafeCommandStore safeStore, Timestamp executeAt, PartialTxn txn)
         Ranges unavailable = safeStore.ranges().unsafeToReadAt(executeAt);
 
         txn.read(safeStore, executeAt).begin((next, throwable) -> {
-            if (throwable != null)
+            // TODO (expected, exceptions): should send exception to client, and consistency handle/propagate locally
+            logger.trace("{}: read failed for {}: {}", txnId, unsafeStore, throwable);
+            synchronized (ReadData.this)
             {
-                // TODO (expected, exceptions): should send exception to client, and consistency handle/propagate locally
-                logger.trace("{}: read failed for {}: {}", txnId, unsafeStore, throwable);
-                node.reply(replyTo, replyContext, ReadNack.Error);
+                if (fail == null)
+                    fail = throwable;
+                else
+                    fail.addSuppressed(throwable);
             }
-            else
-                readComplete(unsafeStore, next, unavailable);
+            readComplete(unsafeStore, next, unavailable);
```

Comment:
if there is a failure do we really want to call `readComplete`?  Feel like we should try to halt right away as the coordinator has already seen our failure

Reply:
I left it calling `readComplete` (exceptionally or otherwise) so that it would eventually send the response once all command stores responded with `fail` containing all the errors that occurred if there are more than one.

I am not married to that decision though. Sending back the first error is fine.

`readComplete` manipulates the `waitingOn` bit set so I left it at that level. `readComplete` does have one override that actually does something different, but since everything always goes through `ack` and `ack` checks for failures I was going to call it good enough.

It might be cleaner to put the `waitingOn` update into `ack` and call `ack` directly. WDYT?

Reply:
ill look closer before replying directly to your comments, but one thing I am thinking about; when we lived with everyone needing to handle exceptions, no one did... this patch is trying to improve that.... so for this to hope all extending classes "do the right thing" might be good much to ask.... but ill reread this code and reflect on your comment

Reply:
The derived classes aren't the ones that send back the errors. `ack` does and that is `private`.

The derived classes can (and do) drop errors as part of their execution such as `FetchRequest.readComplete` which starts a new `AsyncChain` and adds a callback which passes the failure to `uncaughtExceptionHandler`.

I think the interface for listeners, task submission, and callbacks errors is possibly what is wrong. Every execution should occur with a context that specifies what the uncaught error handling should be, but you don't have to provide that context along with the thing you are executing. Or you do, but it's bespoke to that one thing you want to run (parameter to callback) rather than common across a variety of things that might be done to execute a request.

That would help with making it so can't get an error and then drop it or consume it in the wrong place because you would never handle it in the first place unless you specifically provided your own error handler rather than the one in the context. And when you do specify it should still hand off errors you don't want to deal with to the parent error handling context.

Reply:
before this patch this is what I see

* accord.messages.ReadData#read would reply `Error` on the first exception and not call `readComplete`.  This would not update `waitingOn`, but more importantly it would not call `ack`; which means `if (-1 == --waitingOnCount)` will never be true!
* `accord.impl.AbstractFetchCoordinator.FetchRequest#readComplete` makes the read 2 operations:`txn.read()` and `store.maxAppliedFor(...).onSuccess(a -> {this.maxApplied = a; super.readComplete(...)})` (I overly simplified).  Even if a fetch had 2 reads (C* only has 1 atm), then we would still not see the ACK so would never complete the operation....
* `accord.messages.ReadTxnData#readComplete` documents that it should unregister its listener... but because we call `cancel` we will schedule removal of the listener anyways... so thats prob fine.

> I am not married to that decision though. Sending back the first error is fine.

I guess I would ask, what will the coordinator do differently with a list of failures?  If a single one exists we will do the same as if 100 existed...  so by only sending the first we have a smaller msg payload and can start handling the exception faster!  We could detect this case and just log locally that these exceptions are getting ignored....

So, in short, I am in favor of failing fast and not pushing handling exceptions properly to each impl.

Reply:
> I think the interface for listeners, task submission, and callbacks errors is possibly what is wrong. Every execution should occur with a context that specifies what the uncaught error handling should be,

I agree with you on this point... AsyncChain and Futures are best when we "chain" them and centralize the error handling, but when we keep mixing and matching it gets hard to get right...

the common pattern is

```
// this is actually async, but to simplify the code I didn't do that
A value = null;
for store in stores.intersects(request):
  A local = store.process(request.map)
  value = request.reduce(value, local);
request.accept(value)
```

and in this case we try to add more state to make sure all `map` are called and `accept` is called (which is why we use `-1` in `-1 == --waitingOnCount`)...  

We could change this pattern to be

```
List<AsyncChain<A>> chains;
for store in stores.intersects(request):
  chains.add(store.process(request.map));
AsyncChains.reduce(chains, request.reduce)
```

this would allow us to start composing and chaining, rather than stoping chains, and forking new ones in different places...

Now, given how that is a structural change, I would prefer to do that as its own standalone work; adds far too much risk to this patch...